pontoon/sync/sync_project.py

```diff
@@ -0,0 +1,100 @@
+import logging
+
+from django.utils import timezone
+from notifications.signals import notify
+
+from pontoon.base.models import ChangedEntityLocale, Locale, Project, User
+from pontoon.sync.checkouts import get_checkouts
+from pontoon.sync.models import ProjectSyncLog, RepositorySyncLog, SyncLog
+from pontoon.sync.paths import get_paths
+from pontoon.sync.sync_entities_from_repo import sync_entities_from_repo
+from pontoon.sync.sync_translations_from_repo import sync_translations_from_repo
+from pontoon.sync.sync_translations_to_repo import sync_translations_to_repo
+
+log = logging.getLogger(__name__)
+
+
+def sync_project(
+    project_pk: int,
+    sync_log_pk: int,
+    pull: bool = True,
+    commit: bool = True,
+    force: bool = False,
+):
+    try:
+        project = Project.objects.get(pk=project_pk)
+        sync_log = SyncLog.objects.get(pk=sync_log_pk)
+    except Project.DoesNotExist:
+        log.error(f"Could not sync project with pk={project_pk}, not found.")
+        raise
+    except SyncLog.DoesNotExist:
+        log.error(
+            f"Could not sync project {project.slug}, log with pk={sync_log_pk} not found."
+        )
+        raise
+
+    # Mark "now" at the start of sync to avoid messing with
+    # translations submitted during sync.
+    now = timezone.now()
+
+    log_prefix = f"[{project.slug}]"
+    log.info(f"{log_prefix} Sync start")
+    project_sync_log = ProjectSyncLog.objects.create(
+        sync_log=sync_log, project=project, start_time=now
+    )
+
+    try:
+        checkouts = get_checkouts(project, pull)
+        paths = get_paths(project, checkouts)
+    except Exception as e:
+        log.error(f"{log_prefix} {e}")
+        project_sync_log.skip()
+        raise e
+
+    locale_map: dict[str, Locale] = {
+        lc.code: lc for lc in project.locales.order_by("code")
+    }
+    paths.locales = list(locale_map.keys())
```

Comment:
Atm `get_paths()` is only looking at what's in the repo, and that's going to have some set of locales. This is here because it's info we're getting from the db, which is then used not just in `paths` but elsewhere as well.